cityofaustin/atd-moped
# Add `substantial_completion_date` to project and component views

## Associated issues
- https://github.com/cityofaustin/atd-data-tech/issues/15173
- https://github.com/cityofaustin/atd-data-tech/issues/15242

## Testing
**URL to test:**  Local

You are testing to make sure the `substantial_completion_date` meets these requirements:
1. At the project level, the date is derived from the earliest `moped_proj_phases` `phase_start` or `phase_end` where the date is _confirmed_ and the simple phase name is `complete`.
2. At the component level, use the `completion_date` if not null, otherwise use the `substantial_completion_date` from the `project_list_view`.

As a reminder, the field `is_phase_start_confirmed` and `is_phase_end_confirmed` are both set via DB trigger based on whether or not the date is in the future. This behavior will change once it's possible to [edit project phases with a form](https://github.com/cityofaustin/atd-moped/pull/1195).

1. Create a project and add a **Post-construction** phase to the project. Set the phase start date to yesterday.
2. Use your SQL client to view the project's `substantial_completion_date` in the `project_list_view`. It should match the post-construction phase start date (although we have known [timezone issue](https://github.com/cityofaustin/atd-data-tech/issues/14342) with phase dates)
4. Add a component to your project and use the component form to set the component's completion date to tomorrow (you will need to flip the "Use component phase" toggle to see the date input)
5. Check your component's `substantial_completion_date` in the `component_arcgis_online_view`. It should match the component's completion date.
6. Clear the component's completion date and check the `component_arcgis_online_view` again. Now the `substantial_completion_date` should match the start date of the project's **Post construction phase**
7. Remove the start date from your project's **Post construction phase** and set the phase end date to tomorrow.
8. Check the `substantial_completion_date` in the `project_list_view` and the `component_arcgis_online_view` — in both cases  `substantial_completion_date` should be `null`.


Currently, it's not possible edit any date's confirmation status. This is going to be easier to test this once you can [edit a date's confirmation status](https://github.com/cityofaustin/atd-moped/pull/1195).


---
#### Ship list
